ColorPalette: Add an accessible color picker based on ChromePicker

ryelle opened the pull request:
This PR attempts to fork out the ChromePicker from `react-color` to make accessibility-related improvements. I've introduced a new component `ColorPicker` in `components/src/color-picker`, which is visually very similar to the existing color picker. All components were pulled from [react-color](https://github.com/casesandberg/react-color/tree/master/src/components/chrome), then updated to follow GB code standards, then updated with the accessibility improvements. cc @tofumatt 

**Current `ChromePicker` (on master)**

![chromepicker-master](https://user-images.githubusercontent.com/541093/46886947-57b18480-ce2a-11e8-9637-80912c19d029.png)

**New ColorPicker (this PR)**

| Hex input | RGB input | With alpha |
| --- | --- | --- |
| ![color-picker-hex](https://user-images.githubusercontent.com/541093/46886959-613aec80-ce2a-11e8-9539-770d07e97e78.png) | ![color-picker-rgb](https://user-images.githubusercontent.com/541093/46886960-613aec80-ce2a-11e8-8b66-25ab1757405c.png) | ![color-picker-w-alpha](https://user-images.githubusercontent.com/541093/46886961-613aec80-ce2a-11e8-929b-384667c48a5a.png) |

I've kept the alpha functionality in ChromePicker, as it was referenced in #4726 that we might want to enable this.

The accessibility improvements I made are:

- Text inputs are `type=number` so that arrow keys can increase/decrease values
- Text inputs are correctly labelled, and grouped in fieldsets for RGB/HSL
- Toggle input type button is labelled and a `button` so it supports keyboard interaction
- Hue & Alpha bars have been implemented [as `slider`s](https://www.w3.org/TR/wai-aria-practices/#slider), and support arrow keys to increase/decrease values (along with shift+arrows, page up, page down, home, and end)
- Saturation has 2 axis of data: up and down change brightness, while left + right change saturation. These also have keyboard support, but there isn't a matching aria role. Instead I've added a `aria-describedby` attribute to say "Use your arrow keys to change the base color. Move up to lighten the color, down to darken, left to increase saturation, and right to decrease saturation."

I've tested this on Mac/Chrome, Mac/Safari, Mac/Safari/VoiceOver, Windows/Firefox/NVDA. I'm personally having some trouble with the sliders when using a screen reader, but I have the same problems on the examples, so that might be user error (since I don't use a screen reader daily). I'm also open to the idea of changing the saturation interaction if there is a better role/pattern we can use. 

The toggle handles themselves (the little circles that show position over the color graphs) don't have a defined focus style at the moment, so getting a designer's feedback there would be great 🙂 Same for the text inputs, too.

**To test yourself:**

- Open the editor
- Select a block with color settings, for example paragraph
- In the block settings, open "Color settings"
- Click on the multicolored "custom color picker" button
- Interact with the color picker: 
  - enter a hex color
  - attempt to change the color using just your keyboard (you can tab forward or shift+tab backwards between controls)
  - pick a color using your mouse (no regressions 🙂 )
  - toggle to RGB input & enter a color in that format
  - try inputting invalid color values

## Checklist:
- [ ] My code is tested. _I haven't added tests yet, but I'm happy to add that if this approach is valid._
- [x] My code follows the WordPress code style
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

This PR doesn't remove the `react-color` dependency, so we'll want to follow up with that (or I can add it here -- wasn't sure of the process).

ryelle committed:
Add color picker component

ryelle committed:
Add color picker styles to build

ryelle committed:
Update to buttons for focus handles

ryelle committed:
Update styles

ryelle committed:
Update class name

ryelle committed:
Add color-picker component to exports

ryelle committed:
Update name to ColorPicker

ryelle committed:
Add labels to color input groups

ryelle committed:
Fix describedby prop name

ryelle committed:
Update view to use `getDerivedStateFromProps`

ryelle committed:
Add “Input” component to handle intermediate state when entering values

Specifically needed for hex colors, which can be invalid while they’re being entered

ryelle committed:
Add space around text inputs

dimadin commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+				<div className="components-color-picker__inputs-toggle">
+					<IconButton
+						icon="arrow-down-alt2"
+						label="Toggle input type"
```
Should this label be marked for translation?

mtias commented:
This is great, thank you @ryelle!

ryelle committed:
Translate toggle label

youknowriad reviewed (approved):
Awesome work here. I skimmed through the code (not deeply) and it's looking good to me. 

Let's get rid of `react-color` as it's not used anymore. and yeah test would be great but I'm fine leaving for later to get this in the UI-freeze milestone. I'll defer to @tofumatt for accessibility validation.

youknowriad commented:
(Looks like there's an outdated snapshot causing the unit tests to fail)

mtias commented:
Are some of our color utilities (contrast, etc) duplicating any of the libraries work (`tinycolor2`)? cc @jorgefilipecosta

aduth commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+			return (
+				<fieldset>
+					<legend className="screen-reader-text">
+						{ __( 'Color value in HEX' ) }
```
Minor: Is "HEX" a word that ought be capitalized?

ZebulanStanphill replied:
I'd say no. It is not an acronym like RGB ("red-green-blue"); it is an abbreviation of "hexadecimal".

Actually, why not just say "Color value in hexadecimal"?

tofumatt replied:
"Color value in hexadecimal" works for me.

afercia replied:
When there's only one input field, a fieldset + legend are not so appropriate. They're meant to group a set of controls and give the group a name. It is redundant and potentially confusing for users, as they would hear there's a group when actually there's no group:

<img width="636" alt="screen shot 2018-10-17 at 16 13 22" src="https://user-images.githubusercontent.com/1682452/47093730-13f3bc00-d22a-11e8-9d85-3b3347d3e333.png">

I'd suggest to remove fieldset + legend for the hex color and use "Color value in hexadecimal" as the visible `<label>` element of the field.

nosolosw replied:
Fixed at 8490b97

ryelle commented on packages/components/src/color-picker/alpha.js:
```diff
@@ -0,0 +1,142 @@
+					style={ gradient }
+				/>
+				{ /* eslint-disable jsx-a11y/no-static-element-interactions */ }
+				<div
```
This here was done to make the forking of react-color's ChromePicker as 1:1 as possible. Without looking into it more, I'm not sure if we can (as easily) style the gradient background on a range `input`.

ryelle commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+		this.handleKeyDown = this.handleKeyDown.bind( this );
+	}
+
+	componentWillReceiveProps( nextProps ) {
```
I saw that while I was working, but I'm not sure how to best handle this - `getDerivedStateFromProps` doesn't give me access to previous props, and I only want the state to update if the prop has changed (as that indicates a change from interacting with the sliders).

tofumatt replied:
Would `getSnapshotBeforeUpdate` work? It has `prevProps` as the first argument. It seems to fire before DOM mutation and reading the docs that implies `this.props` inside that function would be the equivalent to `nextProps` here but I'm not sure 🤷‍♂️

ryelle committed:
Use KeyboardShortcuts component to avoid separate instances of Mousetrap

tofumatt commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+
+			this.props.onChange( {
+				h: data.h || this.props.hsl.h,
+				s: Number( ( data.s && data.s ) || this.props.hsl.s ),
```
It can be, so this should account for that.

nosolosw replied:
Also to note: these values are taken from the `input type=number` component and can be void. The same for the RGB code above.

nosolosw replied:
`data.s` will always be a string. It can be `"0"` which is a truthy value. So, the only case where the `this.props.s` is used is when `data.s` is blank.

tofumatt reviewed (changes requested):
Sorry it took me so long to get to this!

So I took this for a spin and the keyboard accessibility is next-level awesome. It's intuitive and I dig it. I'm going to bed soon so I'm gonna run through it with screen-readers tomorrow, but as noted the lack of focus styles on the sliders is the main visual issue I'm having with this. We definitely need those, from maybe @karmatosed or @sarahmonster? Failing that we can just hack _something_ together because it's not clear when I have one focused, but the colour picker itself works a treat.

I'll give a run through with a screenreader tomorrow, but for now I'd say we need a focus style on the slider dots.

tofumatt reviewed (changes requested):
I tried this out with Firefox + NVDA and it largely worked pretty well, there are certainly plenty of little tweaks we could make, but I want to focus on the major ones for now. I've left inline comments everywhere I can so it's easier to address things.

One issue is that keyboard focus isn't trapped inside the modal, at least in Firefox + NVDA. Here's a video (contains audio from NVDA): https://www.dropbox.com/s/7ipcjfhn3gs9gdh/Screen%20Recording%202018-10-16%20at%2015.17.10.mov?dl=0

That's after tabbing to the control. You can here the `aria-valuenow` non-integer issue in there too.

### Initial focus

One last issue is when I press <kbd>Enter</kbd> on the custom colour picker I don't have keyboard focus in the colour picker and can't control it. Or actually this could be related to the keyboard up/down/left/right not working in Firefox for the Hue/Lightness picker. It's hard to describe but I can send a screencast if needed, just let me know if I'm not being clear.

Overall though this is great. We should try to keep it simple for now and focus on what I've mentioned, I'll see if there's anything else after but this is so much better than before.

sarahmonster commented:
Here's a suggested solution for the focus styling:

![2018-10-16 20 13 06](https://user-images.githubusercontent.com/376315/47041178-0a157e80-d180-11e8-877e-9208610dc640.gif)

I'd also suggest applying a blue-outline focus styling to the "Toggle input type" drop-down button, which should maybe say "Change input type" rather than "toggle" as well, since toggle tends to refer to a binary switch and is represented by physical toggles in the UI.

tofumatt commented:
Those styles work for me. Can you push them to GitHub or paste them as CSS here so we can apply them?

sarahmonster committed:
Add focus styles to colourpicker pointers.

tofumatt commented:
To clarify about focus not being trapped/keyboard up/down escaping the modal; here's a GIF of me pressing up/down/left/right (it's not logged on screen, sorry, but I'm doing it) a bunch in the colour picker after focusing it with the keyboard. You can see it's focused thanks to @sarahmonster's focus styles, but it's not moving and eventually you can notice elements behind the popover are being selected.

Do we need to use another method to trap the keyboard in that colour picker? @aduth, I think you know a bunch about our keyboard/focus traps–can you spot what's not working here?

(I'm in Firefox for Windows, btw.)

tofumatt commented on packages/components/src/color-picker/alpha.js:
```diff
@@ -0,0 +1,142 @@
+						role="slider"
+						aria-valuemax="1"
+						aria-valuemin="0"
+						aria-valuenow={ rgb.a }
```
Ah, right on. Since `aria-valuetext` is an alternative, even if not super-supported we could use that for the integer value and have `aria-valuenow` as the less-friendly-but-usable decimal number.

afercia replied:
> NVDA read off a lot of decimal places

Yeah (see below(, same for JAWS. They literally announce all those decimals:

![screenshot 105](https://user-images.githubusercontent.com/1682452/47099501-b82f3000-d235-11e8-99dd-59d9c962aaca.png)

afercia replied:
> Ah, right on. Since `aria-valuetext` is an alternative, even if not super-supported we could use that for the integer value and have `aria-valuenow` as the less-friendly-but-usable decimal number.

I guess we can test it with the major browser / screen reader combos and if it has no good support revert to using `aria-valuenow` with the friendly value.

afercia commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+				<div className="components-color-picker__inputs-toggle">
+					<IconButton
+						icon="arrow-down-alt2"
+						label={ __( 'Toggle input type' ) }
```
I'd consider to clarify this label. It certainly toggles the input type, but the verbiage seems a bit too technical to me. I'd suggest to use a text that clearly explains what it does, something along the lines of: "Toggle color format".

tofumatt replied:
Oops, yeah, @sarahmonster pointed that out to me in-person and I forgot. It's really not even a "toggle" between that implies a binary change; this actually goes through more than two colour selection options.

tofumatt replied:
So maybe something like "Change color picker format"?

sarahmonster replied:
Yes, let's please swap out "toggle" for "change" here! (I think I mentioned this in overall comments.)

"Change color mode" might be more accurate here, although keeping in mind it's likely that a majority of users won't know what colour modes are.

sarahmonster replied:
```suggestion
						label={ __( 'Change color mode' ) }
```

nosolosw replied:
Fixed at 4f82bae

afercia commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+						icon="arrow-down-alt2"
+						label={ __( 'Toggle input type' ) }
+						onClick={ this.toggleViews }
+					/>
```
When toggling color format, there's no audible feedback about what just happened. Screen reader users would be forced to explore the UI to get an idea of what just changed. A simple `speak` message could help and give an useful feedback.

tofumatt replied:
Added in bbfe56a3bc2c24ab0ca1f5ee564d49013514ae10.

afercia commented on packages/components/src/color-picker/saturation.js:
```diff
@@ -0,0 +1,151 @@
+					aria-describedby={ `color-picker-saturation-${ instanceId }` }
+					className="components-color-picker__saturation-pointer"
+					style={ pointerLocation }
+				/>
```
Initial focus goes to this button and it gets announced as:
"Saturation & lightness value button"

![screenshot 104](https://user-images.githubusercontent.com/1682452/47096275-1efd1b00-d22f-11e8-92e0-f312f28056df.png)

I'm not sure a `button` is the best element to use, because its role suggests it can be pressed, not "moved" on a canvas. In combination with its label `Saturation & lightness value button` I'm not sure it's clear enough what this control is about for users. How this could be improved? I can't think of an ARIA role that is suitable for this control. However, maybe we can try to improve its aria-label. Considering initial focus is set on this button, instead of 

`Saturation & lightness value`

I'd propose to simplify and use:

`Pick a color` - followed by the detailed aria-description - `Use your arrow keys to change ...` 

A combination of label and description along those lines could help.

sarahmonster replied:
"Color" might be confusing here, since this area is for tweaking the saturation and lightness, and there's a second slider below that allow for selecting more of what people think of as "colour" (the hue, that is). 

It's not *technically* correct, per se, but I think we could get away with saying "Choose a shade" here and have people understand generally what that refers to.

ryelle replied:
I've updated the label to read "Choose a shade" in 51f12019004992a3f2618880e3bb773fd75ceb93 . I haven't moved the components though, I think for now we should leave the sliders as they are (and maybe follow up with another PR if it still feels too awkward).

ryelle commented:
Thanks everyone for all the feedback 🙂 just a status update: I've been at a conference the last few days, but should have time later today to apply these changes.

afercia commented:
@ryelle thanks so much for working on this. I've left some comments.

Additionally, seems the saturation "dot" button and the slider buttons don't work as expected with Safari + VoiceOver. On Windows, NVDA and JAWS seem not affected. [As discussed on Slack](https://wordpress.slack.com/archives/C02RP4X03/p1539793296000100), an option could be attaching a keydown event on the buttons and prevent the default (but also allowing tabbing away). That fixed it for me. Although screen readers tend to intercept keystrokes and events might not behave as expected, I wonder why there's the need to prevent a `keydown` default in this specific case ¯\_(ツ)_/¯

Minor:
there are a few `overflow: hidden` values, including one from the popover component, that prevent the "dot" focus style to be fully visible when the dot is close to the edges of its container:

<img width="307" alt="screen shot 2018-10-17 at 16 04 49" src="https://user-images.githubusercontent.com/1682452/47102443-14e21900-d23d-11e8-9223-4af8056048ca.png">

afercia commented on packages/components/src/color-picker/alpha.js:
```diff
@@ -0,0 +1,142 @@
+					onTouchStart={ this.handleChange }>
+					<button
+						role="slider"
+						aria-valuemax="1"
```
Not sure a button is a proper element for a `role="slider"`. We're basically overriding the native semantics of the button and it's against the W3C recommendations. See https://www.w3.org/TR/using-aria/#notes2

Also, the ARIA [slider implementation details](https://www.w3.org/TR/wai-aria-practices-1.1/#slider) and the examples ([example 1](https://www.w3.org/TR/wai-aria-practices-1.1/examples/slider/slider-1.html) and [example 2](https://www.w3.org/TR/wai-aria-practices-1.1/examples/slider/slider-2.html)) simply use a focusable div. The proper semantics is given by the ARIA role. This applies also to the other slider.

ryelle replied:
I needed to use a natively focusable element like a button for a screen reader (I think this was VoiceOver) to properly pick up on key events - but that issue is fixed with the key events fix in 5fddf98af8d9cffa3b85b705356ef6427e0cdd7b, so I've updated this to a focusable div in 1ea4e2a3486e103a72e6b05a5ae32fd5b34a48a0.

afercia commented:
Forgot to mention I'd propose to consider to expand the input fields labels:
`r`, `g`, `b`, `a`

and 
`h`, `s`, `l`, `a`

Having just one letter as accessible name of a control is not so ideal for all users, and is problematic for assistive technology users. On the other hand, I understand there's little space available. Maybe some CSS magic could help.

gziolo commented on packages/components/src/color-palette/index.js:
```diff
@@ -2,7 +2,6 @@
  * External dependencies
  */
 import classnames from 'classnames';
-import { ChromePicker } from 'react-color';
```
We need to remove `react-color` from `package.json` file as it is no longer used:

https://github.com/WordPress/gutenberg/blob/master/packages/components/package.json#L45

nosolosw replied:
fixed at 3c4686e

gziolo commented on packages/components/src/color-picker/index.js:
```diff
@@ -0,0 +1,93 @@
+import Saturation from './saturation';
+import { colorToState, simpleCheckForValidColor } from './utils';
+
+export default class extends Component {
```
This component should have name `ColorPicker` so it would be possible to find it in React dev tools.

```js
export default class ColorPicker extends Component {
```

nosolosw replied:
Fixed at d50deda

gziolo commented on packages/components/src/color-picker/utils.js:
```diff
@@ -0,0 +1,185 @@
+ * External dependencies
+ */
+import { each } from 'lodash';
+import tinycolor from 'tinycolor2';
```
`tinycolor2` needs to be added to dependencies in `package.json` file:
https://github.com/WordPress/gutenberg/blob/master/packages/components/package.json#L22

nosolosw replied:
Fixed at b8b9f37

gziolo commented on packages/components/src/index.js:
```diff
@@ -9,6 +9,7 @@ export { default as CheckboxControl } from './checkbox-control';
 export { default as ClipboardButton } from './clipboard-button';
 export { default as ColorIndicator } from './color-indicator';
 export { default as ColorPalette } from './color-palette';
+export { default as ColorPicker } from './color-picker';
```
We are about to add a new component so we should update changelog file and list it as a new feature and propose minor version bump, see: https://github.com/WordPress/gutenberg/blob/master/CONTRIBUTING.md#maintaining-changelogs and file to be updated: https://github.com/WordPress/gutenberg/blob/master/packages/components/CHANGELOG.md.

gziolo commented on packages/components/src/color-picker/index.js:
```diff
@@ -0,0 +1,93 @@
+/**
```
Can we add a corresponding README file with a simple usage example so it could be exposed in both Gutenberg Handbook and Calypso DevDocs?
A similar example: https://github.com/WordPress/gutenberg/blob/master/packages/components/src/color-palette/README.md.

gziolo commented:
> All components were pulled from react-color, then updated to follow GB code standards, then updated with the accessibility improvements.

We need to include credits in files that were ported from `react-color`. See one of the previous cases:
https://github.com/WordPress/gutenberg/blob/master/packages/element/src/serialize.js#L1-L26

oandregal committed:
Remove react-color dependency

oandregal committed:
Add class name to better find the component in the devtools

oandregal committed:
Add tinycolor2 dependency

oandregal committed:
Add reference to react-color license & copyright

oandregal committed:
Better label

oandregal committed:
Update snapshot

oandregal committed:
Decouple valueKey and label

so we can use semantic labels.

nosolosw commented on packages/components/src/color-picker/hue.js:
```diff
@@ -0,0 +1,129 @@
+						role="slider"
+						aria-valuemax="1"
+						aria-valuemin="359"
+						aria-valuenow={ hsl.h }
```
The decimal values are derived from [`colorToState` utility](https://github.com/WordPress/gutenberg/pull/10564/files#diff-fb44e1885b236e2d9a941fb71810f963R41). That point seems to be the only place where those are calculated, so we may be able to round them there. Looking.

tofumatt commented on packages/components/src/color-picker/inputs.js:
```diff
@@ -0,0 +1,287 @@
+import { isValidHex } from './utils';
+import TextControl from '../text-control';
+
+/* Wrapper for TextControl, only used to handle intermediate state while typing. */
```
We should, but we should file a follow-up issue for that… I will set a reminder to do it; trying to fix issues in here first. 😄

tofumatt replied:
Filed #10974 😄

tofumatt committed:
Use createRef

oandregal committed:
Normalize HSV / HSL values

H: 0 - 260
S, L, V: 0 - 100

ryelle committed:
Omit `valueKey` from being passed to input

ryelle committed:
Use createRef in all components

ryelle commented on packages/components/src/color-picker/hue.js:
```diff
@@ -0,0 +1,129 @@
+						aria-valuemin="359"
+						aria-valuenow={ hsl.h }
+						aria-orientation="horizontal"
+						aria-label={ __( 'Hue value in degrees, from 0 to 359.' ) }
```
Is a description necessary? <kbd>up</kbd>/<kbd>down</kbd> will also work for the 1-dimensional sliders (hue/alpha), and follows the expected behavior of a slider.

ryelle committed:
Update saturation handle label

ryelle committed:
Prevent key events on slider handles

Fixes the issue where VoiceOver grabs arrow-key input as navigation

ryelle committed:
Use a focusable div for slider handle

ryelle committed:
Apply focus style to alpha handle

tofumatt commented:
I'm not able to move the slider using a keyboard in Windows. Were you able to move the slider using the keyboard in Windows? Maybe it's just a bug in my VMWare setup. Here's me trying; note the focus leaves the colour picker:

![2018-10-19 02 46 55](https://user-images.githubusercontent.com/90871/47193229-c2425300-d349-11e8-9839-98bd76c02b1d.gif)

tofumatt committed:
Add description to hue controls

tofumatt committed:
Add speak commands for view change

tofumatt commented:
> Forgot to mention I'd propose to consider to expand the input fields labels:
> r, g, b, a
> 
> and
> h, s, l, a
> 
> Having just one letter as accessible name of a control is not so ideal for all users, and is problematic for assistive technology users. On the other hand, I understand there's little space available. Maybe some CSS magic could help.

This is a really good point, but the properly display these they'd need localisation, which is difficult for us to handle with string interpolation + React at the moment (see #9846). Unless we should always display H/S/L but have screen-reader elements in div alongside/as labels for each letter that explained the labels in a localised way… maybe that could work?

I that that's worth doing but could be done after this gets into the UI freeze. There are a fair few follow-up issues to be extracted from this PR, but they're better-handled as iterations as this PR is already getting pretty tough to manage because GitHub hides a lot of comments.

I will go through after it's merged and file follow-ups for anything we *don't* address here, but it's important we get this in, at least as a basis for iteration, for the UI freeze.

tofumatt reviewed (commented):
It looks like all that remains for this to be way better than what we have, an excellent basis for iteration, and something that covers most of the actual visual UI changes we'd want to make here, is to fix the package issues @gziolo mentioned. Let me know if there's anything I can do to help that part of the PR or to clarify anything else I might've missed. I tried to go through all the comments but it can sometimes be tough to tell in these big PRs what's left to do; just the way GitHub handles them 🤷‍♂️

If this is working in Windows for people I need to update my VMWare setup. It also means I think this is good to go and get merged.

If someone can confirm this works well for them on Windows and the new component pieces get added as per @gziolo's request I say we get this merged 👍

Thanks so much for all your work on this @ryelle; it's an epic patch and I am really pleased to see it. It was the big UI component in the Merge: Accessibility milestone I wasn't sure we'd manage to close; happy you proved my doubts wrong 😄

oandregal committed:
Update snapshot

afercia commented:
> If someone can confirm this works well for them on Windows

Tested and it doesn't work with screen readers on Windows. Not sure why it apparently worked when previously tested. Regardless:
- screen readers on Windows intercept keystrokes when they're in their default "browse mode"
- when landing on the shade button, screen readers don't automatically switch to "forms mode" because it's a button
- on the other sliders, screen readers do change mode because they're `role="slider"`
- solution: the shade needs a way to make screen readers stop intercepting keys
- a `role="slider"` would make it but it's not an appropriate role
- the only way I can think of is setting a `role="application"` on the saturation `<div>` wrapper (the one with `ref={ this.container }`
- at that point, the div will also need to be labelled with something like `aria-label="Shade picker"` (will leave choosing a better wording to native English speakers
- please consider `role="application"` must be used with caution and very sparingly, as it resets any native semantics and has other serious drawbacks; please always ask to the a11y team before using it

I've tested `role="application"` on various combos on Windows:
- NVDA Firefox
- NVDA IE11
- JAWS Chrome
- JAWS Firefox
- JAWS IE11

And it worked for me. The `aria-label` would be announced before the button label and description:

![screenshot 107](https://user-images.githubusercontent.com/1682452/47209045-aeb9db00-d38f-11e8-97ec-54d75b87c978.png)

Thanks so much for all the work and testing done on this component.

afercia commented on packages/components/src/color-picker/saturation.js:
```diff
@@ -0,0 +1,186 @@
+		/* eslint-disable jsx-a11y/no-static-element-interactions */
+		return (
+			<KeyboardShortcuts shortcuts={ shortcuts }>
+				<div
```
The `role="application"` I've suggested should be set here, together with a meaningful `aria-label`.

tofumatt reviewed (changes requested):
I think the chief thing here is to get the keyboard inputs working on Windows.

I don't recall them working for me with or without NVDA on.

tofumatt commented:
@gziolo does the commit stuff look better now? I think I cleaned it up.

gziolo commented:
![screen shot 2018-10-19 at 11 49 55](https://user-images.githubusercontent.com/699132/47211167-20485800-d395-11e8-9e69-d1d2dfba8e29.png)

Oh yeah, much better, it was +8 000 + lines before ;)

oandregal committed:
Remove test as the component API surface changed

We have done some changes to the ChromePicker so this test
can no longer be replicated.

gziolo commented:
Can someone apply the following patch

```diff
diff --git a/packages/components/CHANGELOG.md b/packages/components/CHANGELOG.md
index 2665f8e62..106fd58c4 100644
--- a/packages/components/CHANGELOG.md
+++ b/packages/components/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## 4.2.0 (Unreleased)
 
+### New Feature
+
+- Added a new `ColorPicker` component ([#10564](https://github.com/WordPress/gutenberg/pull/10564)).
+
 ### Deprecation
 
 - `wp.components.PanelColor` has been deprecated in favor of `wp.editor.PanelColorSettings`.
diff --git a/packages/components/src/color-picker/README.md b/packages/components/src/color-picker/README.md
new file mode 100644
index 000000000..d2b1307a0
--- /dev/null
+++ b/packages/components/src/color-picker/README.md
@@ -0,0 +1,23 @@
+# ColorPicker
+
+Accessible color picker.
+
+_Parts of the source code were derived and modified from [react-color](https://github.com/casesandberg/react-color/), released under the MIT license._
+
+## Usage
+```jsx
+import { ColorPicker } from '@wordpress/components';
+import { withState } from '@wordpress/compose';
+
+const MyColorPicker = withState( {
+	color: '#f00',
+} )( ( { color, setState } ) => {
+	return (
+		<ColorPicker
+			color={ color }
+			onChangeComplete={ ( value ) => setState( value.hex ) }
+			disableAlpha
+		/>
+	);
+} );
+```
diff --git a/packages/components/src/color-picker/test/__snapshots__/index.js.snap b/packages/components/src/color-picker/test/__snapshots__/index.js.snap
new file mode 100644
index 000000000..c143801f9
--- /dev/null
+++ b/packages/components/src/color-picker/test/__snapshots__/index.js.snap
@@ -0,0 +1,87 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`ColorPicker should render color picker 1`] = `
+<div
+  className="components-color-picker is-alpha-disabled"
+>
+  <div
+    className="components-color-picker__saturation"
+  >
+    <WithInstanceId(Saturation)
+      hsl={
+        Object {
+          "a": 1,
+          "h": 0,
+          "l": 100,
+          "s": 0,
+        }
+      }
+      hsv={
+        Object {
+          "a": 1,
+          "h": 0,
+          "s": 0,
+          "v": 100,
+        }
+      }
+      onChange={[Function]}
+    />
+  </div>
+  <div
+    className="components-color-picker__body"
+  >
+    <div
+      className="components-color-picker__controls"
+    >
+      <div
+        className="components-color-picker__swatch"
+      >
+        <div
+          className="components-color-picker__active"
+          style={
+            Object {
+              "backgroundColor": "rgb(255, 255, 255)",
+            }
+          }
+        />
+      </div>
+      <div
+        className="components-color-picker__toggles"
+      >
+        <WithInstanceId(Hue)
+          hsl={
+            Object {
+              "a": 1,
+              "h": 0,
+              "l": 100,
+              "s": 0,
+            }
+          }
+          onChange={[Function]}
+        />
+      </div>
+    </div>
+    <Inputs
+      disableAlpha={true}
+      hex="#ffffff"
+      hsl={
+        Object {
+          "a": 1,
+          "h": 0,
+          "l": 100,
+          "s": 0,
+        }
+      }
+      onChange={[Function]}
+      rgb={
+        Object {
+          "a": 1,
+          "b": 255,
+          "g": 255,
+          "r": 255,
+        }
+      }
+    />
+  </div>
+</div>
+`;
diff --git a/packages/components/src/color-picker/test/index.js b/packages/components/src/color-picker/test/index.js
new file mode 100644
index 000000000..5510da6e1
--- /dev/null
+++ b/packages/components/src/color-picker/test/index.js
@@ -0,0 +1,26 @@
+/**
+ * External dependencies
+ */
+import ShallowRenderer from 'react-test-renderer/shallow';
+
+/**
+ * Internal dependencies
+ */
+import ColorPicker from '../';
+
+describe( 'ColorPicker', () => {
+	test( 'should render color picker', () => {
+		const color = '#fff';
+
+		const renderer = new ShallowRenderer();
+		renderer.render(
+			<ColorPicker
+				color={ color }
+				onChangeComplete={ () => {} }
+				disableAlpha
+			/>
+		);
+
+		expect( renderer.getRenderOutput() ).toMatchSnapshot();
+	} );
+} );
```

gziolo committed:
Add README, tests and update Changelog

gziolo commented:
I no longer see changes from https://github.com/WordPress/gutenberg/pull/10564/commits/3c4686e5ae637b71a6045bdd2df9c3d7ebce0a5b. We should ensure that merging with master didn't remove other commits.

oandregal committed:
Remove react-color dependency

gziolo reviewed (approved):
Changes for packages look good, thanks @nosolosw for help 👍

afercia commented:
@gziolo @tofumatt please consider this doesn't work with screen readers on Window yet so I'd suggest to not merge and implement the solution suggested above.

tofumatt commented:
The previous colour picker was entirely inaccessible to keyboard users and from my tests to screenreaders too…

I really want this to fix everything right away but it's a great starting point and something we can iterate on, I think even after the UI freeze. At the very least it is _keyboard_ accessible now, which is an improvement (notwithstanding Windows issues, which are weird…).

youknowriad commented:
So the choices we have:

 1- Ship this as an improvement in 4.1 and iterate on it for 4.2 for the remaining screen readers
 2- Don't ship the improvement in 4.1 and wait for 4.2.

I vote 1 :)

tofumatt commented:
My main issue is the keyboard nav isn't even working in Windows. But still: it's an improvement even if flawed.

I'm fine with shipping and improving. I'm just doing some quick run-throughs locally. This is the last patch blocking 4.1 as I understand it.

Can we say I get it passing and try a few things, and at 2:20pm/14h20 whatever we have we ship and iterate on? 😄

youknowriad commented:
@tofumatt WFM

tofumatt committed:
Add docs manifest file

tofumatt reviewed (approved):
For me, in Windows, the saturation/lightness control does not work with my keyboard. Focus is changed after keyboard input to things entirely outside the Colour Palette.

That said:

* this works on Mac fine
* keyboard users can use all of the controls *except* the saturation/lightness control (before they could use none)
* we use `speak()` when the modes change

This is a starting point and should be improved upon going forward, but I want to get this merged for a few reasons:

* future updates should not impact visual UI at all and the rest of the UI minimally (basically we just need to fix keyboard on Windows and improve the UI for screen reader users 😄)
* this is on a fork rather than a branch and it's getting a bit difficult to deal with; we've already had some `git` user error that needed repair and I'd like to avoid future issues

I'll go through this and file the appropriate follow-up issues (or just make a PR if possible) that will land in 4.2.

Thanks all for your efforts here, this is way better. 👍

tofumatt committed:
Merge branch 'master' into add/a11y-color-picker

tofumatt commented:
I'm going to file some follow-up issues here soon and link to them from here. We're just getting started! 😄

afercia commented:
> For me, in Windows, the saturation/lightness control does not work with my keyboard

Weird, for me it works. What browser / screen reader combinations are you testing with? please specify the versions.

tofumatt commented:
I would be happy if it's just me. I'm on a Windows 10, x64 VMWare image (from Modern.IE) running:

* Firefox 62.0.3 (64-bit)
* Edge 42.17134.1.0
* latest Chrome

All whilst running NVDA 2018.3.2 and not running any screenreader software.

It happens in all of them, so maybe it *is* a VM bug...

afercia commented:
I guess It's really your VM. I've re-tested on Windows 10 without a screen reader running:
- Firefox latest
- Chrome latest
- Edge
- IE 11
they all work.

With a screen reader (NVDA, JAWS) running no, the saturation/lightness control does not work with a keyboard. It needs the fix suggested previously. Will prepare a quick PR.
